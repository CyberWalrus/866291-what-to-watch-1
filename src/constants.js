const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;
const NUMBER_FILM_GENRE = 4;
const SERVER_URL = `https://es31-server.appspot.com/wtw`;
const SERVER_URL_USER = `https://es31-server.appspot.com/`;
const LINK_STYLE = `inherit`;
const MAIN_FILM_ID = 1;
const PADDING_VIDEO = 25;
const REVIEW_MESSAGE = `OK`;

const ratingRadioValues = [`1`, `2`, `3`, `4`, `5`];

const RatingName = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`
};
const RatingInterval = {
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  VERY_GOOD: 10,
  AWESOME: 10
};
const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const BodyOverflow = {
  VISIBLE: `visible`,
  HIDDEN: `hidden`
};
const FavoriteStatus = {
  ADD: 1,
  REMOVE: 0
};
const TextLength = {
  MIN: 50,
  MAX: 400
};
const RatingValue = {
  MIN: 1,
  MAX: 5
};

const DisabledStyle = {
  opacity: `0.4`,
  border: `inherit`,
  cursor: `not-allowed`
};
const OptionsVideoMin = {
  width: `280`,
  height: `175`,
  isMuted: true,
  isLoop: true,
  isControls: false
};
const OptionsVideoFull = {
  width: `100%`,
  height: `100%`,
  isMuted: false,
  isLoop: false,
  isControls: false
};

export {
  GENRE_DEFOULT,
  NUMBER_FILM,
  SERVER_URL,
  SERVER_URL_USER,
  NUMBER_FILM_GENRE,
  LINK_STYLE,
  PADDING_VIDEO,
  REVIEW_MESSAGE,
  MAIN_FILM_ID,
  ratingRadioValues,
  RatingName,
  RatingInterval,
  FilmRoute,
  OptionsVideoMin,
  OptionsVideoFull,
  BodyOverflow,
  FavoriteStatus,
  DisabledStyle,
  TextLength,
  RatingValue
};
