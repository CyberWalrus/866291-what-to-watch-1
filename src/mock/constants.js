const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;

export {FilmRoute, GENRE_DEFOULT, NUMBER_FILM};

