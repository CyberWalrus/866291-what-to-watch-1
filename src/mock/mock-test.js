const FILMS = [
  {
    bgColor: `#977461`,
    description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
    director: `Martin Scorsese`,
    genre: `Thriller`,
    id: 1,
    isFavorite: false,
    pageUrl: `Shutter Island`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: `8,1`,
    ratingLevel: `Very good`,
    released: 2010,
    runTime: 138,
    scoresCount: 1002557,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
    starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
    title: `Shutter Island`
  },
  {
    bgColor: `#F1E9CE`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    pageUrl: `Macbeth`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: `6,6`,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    starrings: [(`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`)],
    title: `Macbeth`
  }
];
const FILM = {
  bgColor: `#977461`,
  description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
  director: `Martin Scorsese`,
  genre: `Thriller`,
  id: 1,
  isFavorite: false,
  pageUrl: `Shutter Island`,
  srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: `8,1`,
  ratingLevel: `Very good`,
  released: 2010,
  runTime: 138,
  scoresCount: 1002557,
  srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
  srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
  srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
  starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
  title: `Shutter Island`
};
const REVIEWS = [
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  },
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  }
];

const GENRES = [
  `All genres`,
  `Thriller`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Action`,
  `Fantasy`
];
export {FILMS, FILM, GENRES, REVIEWS};
