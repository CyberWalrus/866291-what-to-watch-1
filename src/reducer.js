import films from "./mock/films.js";
const intialState = {
  genreFilter: `all`,
  films
};

const ActionCreator = {
  changeGenre: (value) => {
    if (
      [
        `comedy`,
        `crime`,
        `documentary`,
        `drama`,
        `horror`,
        `family`,
        `romance`,
        `sci-fi`,
        `thriller`
      ].find((item) => item === value)
    ) {
      const filmsNew = films.filter((item) => item.genre === value);
      return {
        type: `CHANGE_GENRE`,
        genreNew: value,
        filmsNew
      };
    }
    return {
      type: `RESET`
    };
  }
};

const reducer = (state = intialState, action) => {
  switch (action.type) {
    case `CHANGE_GENRE`:
      return Object.assign({}, state, {
        genreFilter: action.genreNew,
        films: action.filmsNew
      });
    case `RESET`:
      return Object.assign({}, intialState);
  }
  return state;
};

export {ActionCreator, reducer};
