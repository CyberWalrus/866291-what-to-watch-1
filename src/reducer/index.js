import {combineReducers} from "redux";
import {reducer as filter} from "./filter/filter.js";
import {reducer as data} from "./data/data.js";
import NameSpace from "./name-spaces.js";

export default combineReducers({
  [NameSpace.FILTER]: filter,
  [NameSpace.DATA]: data
});
