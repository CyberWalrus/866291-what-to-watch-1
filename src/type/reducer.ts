import {ThunkDispatch as ReduxThunkDispatch, ThunkAction as ReduxThunkAction} from "redux-thunk";
import {AxiosInstance} from "axios";
import {State as StateData, Action as ActionData} from "../store/data/data";
import {State as StateFilter, Action as ActionFilter} from "../store/filter/filter";
import {State as StateUser, Action as ActionUser} from "../store/user/user";
import NameSpace from "../store/name-spaces";

export interface StateApp {
  [NameSpace.DATA]: StateData;
  [NameSpace.FILTER]: StateFilter;
  [NameSpace.USER]: StateUser;
}

export type ActionApp = ActionData | ActionFilter | ActionUser;
export type ThunkDispatch = ReduxThunkDispatch<StateApp, AxiosInstance, ActionApp>;
export type ThunkAction = ReduxThunkAction<
Promise<void>,
StateApp,
AxiosInstance,
ActionApp
>;
