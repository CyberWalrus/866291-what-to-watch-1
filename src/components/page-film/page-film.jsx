import React, {Fragment} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {getFilm} from "../../store/data/selectors.js";
import Footer from "../footer/footer.jsx";
import HiddenIcon from "../hidden-icon/hidden-icon.jsx";
import Header from "../header/header.jsx";
import FilmOverview from "../film-overview/film-overview.jsx";
import FilmDetails from "../film-details/film-details.jsx";
import FilmReview from "../film-review/film-review.jsx";
import FilmNav from "../film-nav/film-nav.jsx";
import FilmButtonDiv from "../film-button-div/film-button-div.jsx";
import {FilmRoute} from "../../mock/constants.js";
import CardList from "../card-list/card-list.jsx";
import VideScreen from "../video-screen/video-screen.jsx";
import withActiveFilm from "../../hocs/with-active-film/with-active-film.js";
import withVideoScreenState from "../../hocs/with-video-screen-state/with-video-screen-state.js";

const CardListActiveFilm = withActiveFilm(CardList);
const VideScreenState = withVideoScreenState(VideScreen);

const PageFilm = ({film, onChangeFilmRoute, route}) => {
  if (film) {
    const {
      title,
      genre,
      released,
      bgColor,
      srcBgImage,
      srcPosterImage,
      rating,
      scoresCount,
      ratingLevel,
      description,
      director,
      starrings,
      runTime,
      id,
      isFavorite
    } = film;
    return (
      <Fragment>
        <HiddenIcon />
        <VideScreenState />
        <section
          className="movie-card movie-card--full"
          style={{backgroundColor: bgColor}}
        >
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={srcBgImage} alt={title} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <Header />

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genre}</span>
                  <span className="movie-card__year">{released}</span>
                </p>
                <FilmButtonDiv
                  id={id}
                  isFavorite={isFavorite}
                  isShowReview={true}
                />
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img
                  src={srcPosterImage}
                  alt={`${title} poster`}
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <FilmNav onChangeFilmRoute={onChangeFilmRoute} route={route} />
                {
                  {
                    [FilmRoute.OVERVIEW]: (
                      <FilmOverview
                        rating={rating}
                        ratingLevel={ratingLevel}
                        scoresCount={scoresCount}
                        description={description}
                        director={director}
                        starrings={starrings}
                      />
                    ),
                    [FilmRoute.DETAILS]: (
                      <FilmDetails
                        released={released}
                        runTime={runTime}
                        genre={genre}
                        director={director}
                        starrings={starrings}
                      />
                    ),
                    [FilmRoute.REVIEWS]: <FilmReview filmId={id} />
                  }[route]
                }
              </div>
            </div>
          </div>
        </section>

        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <CardListActiveFilm genreFilm={genre} filmId={id} />
          </section>
          <Footer />
        </div>
      </Fragment>
    );
  }
  return <div />;
};
PageFilm.propTypes = {
  id: PropTypes.number.isRequired,
  route: PropTypes.string.isRequired,
  onChangeFilmRoute: PropTypes.func.isRequired,
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    bgColor: PropTypes.string.isRequired,
    srcPreviewImage: PropTypes.string.isRequired,
    srcPosterImage: PropTypes.string.isRequired,
    srcBgImage: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    released: PropTypes.number.isRequired,
    srcVideo: PropTypes.string.isRequired,
    rating: PropTypes.string.isRequired,
    scoresCount: PropTypes.number.isRequired,
    ratingLevel: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    starrings: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    runTime: PropTypes.string.isRequired
  })
};
const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    film: getFilm(state, ownProps.id)
  });

export {PageFilm};

export default connect(mapStateToProps)(PageFilm);
