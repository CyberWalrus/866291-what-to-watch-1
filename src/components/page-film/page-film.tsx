import * as React from "react";
import {Fragment} from "react";
import {connect} from "react-redux";
import {getFilm} from "../../store/data/selectors";
import Footer from "../footer/footer";
import HiddenIcon from "../hidden-icon/hidden-icon";
import Header from "../header/header";
import FilmOverview from "../film-overview/film-overview";
import FilmDetails from "../film-details/film-details";
import FilmReview from "../film-review/film-review";
import FilmNav from "../film-nav/film-nav";
import FilmButtonDiv from "../film-button-div/film-button-div";
import {FilmRoute} from "../../constants";
import CardList from "../card-list/card-list";
import VideScreen from "../video-screen/video-screen";
import withActiveFilm from "../../hocs/with-active-film/with-active-film";
import withVideoScreenState from "../../hocs/with-video-screen-state/with-video-screen-state";
import {Film} from "../../type/data";
import {StateApp} from "../../type/reducer";

const CardListActiveFilm = withActiveFilm(CardList);
const VideScreenState = withVideoScreenState(VideScreen);

interface PropsInsert {
  id: number,
  route: string,
  onChangeFilmRoute: () => void,
}
interface PropsState {
  film: Film;
}
type Props = PropsInsert & PropsState;


const PageFilm = ({film, onChangeFilmRoute, route}: Props) => {
  if (film) {
    const {
      title,
      genre,
      released,
      bgColor,
      srcBgImage,
      srcPosterImage,
      rating,
      scoresCount,
      ratingLevel,
      description,
      director,
      starrings,
      runTime,
      id,
      isFavorite
    } = film;
    return (
      <Fragment>
        <HiddenIcon />
        <VideScreenState />
        <section
          className="movie-card movie-card--full"
          style={{backgroundColor: bgColor}}
        >
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={srcBgImage} alt={title} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <Header />

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genre}</span>
                  <span className="movie-card__year">{released}</span>
                </p>
                <FilmButtonDiv
                  id={id}
                  isFavorite={isFavorite}
                  isShowReview={true}
                />
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img
                  src={srcPosterImage}
                  alt={`${title} poster`}
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <FilmNav onChangeFilmRoute={onChangeFilmRoute} route={route} />
                {
                  {
                    [FilmRoute.OVERVIEW]: (
                      <FilmOverview
                        rating={rating}
                        ratingLevel={ratingLevel}
                        scoresCount={scoresCount}
                        description={description}
                        director={director}
                        starrings={starrings}
                      />
                    ),
                    [FilmRoute.DETAILS]: (
                      <FilmDetails
                        released={released}
                        runTime={runTime}
                        genre={genre}
                        director={director}
                        starrings={starrings}
                      />
                    ),
                    [FilmRoute.REVIEWS]: <FilmReview filmId={id} />
                  }[route]
                }
              </div>
            </div>
          </div>
        </section>

        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <CardListActiveFilm genreFilm={genre} filmId={id} />
          </section>
          <Footer />
        </div>
      </Fragment>
    );
  }
  return <div />;
};

const mapStateToProps = (state: StateApp, ownProps: Props): Props =>
  Object.assign({}, ownProps, {
    film: getFilm(state, ownProps.id)
  });

export {PageFilm};

export default connect(mapStateToProps)(PageFilm);
