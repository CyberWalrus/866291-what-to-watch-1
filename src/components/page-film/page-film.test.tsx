import * as React from "react";
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import {shallow} from "enzyme";
import reducer from "../../store";
import {createStore} from "redux";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import {PageFilm} from "./page-film";
import {FILM} from "../../mock/mock-test";
import {FilmRoute} from "../../constants";
Enzyme.configure({adapter: new Adapter()});

it(`PageFilm correctly renders after relaunch`, () => {
  const handleClick = jest.fn();
  const store = createStore(reducer);
  const tree = toJson(
    shallow(
      <Provider store={store}>
        <BrowserRouter>
          <PageFilm
            id={FILM.id}
            route={FilmRoute.OVERVIEW}
            onChangeFilmRoute={handleClick}
            film={FILM}
          />
        </BrowserRouter>
      </Provider>
    )
  );
  expect(tree).toMatchSnapshot();
});
