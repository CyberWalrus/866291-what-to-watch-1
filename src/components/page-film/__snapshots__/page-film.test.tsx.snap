// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageFilm correctly renders after relaunch 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <BrowserRouter>
    <PageFilm
      film={
        Object {
          "bgColor": "#977461",
          "description": "In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.",
          "director": "Martin Scorsese",
          "genre": "Thriller",
          "id": 1,
          "isFavorite": false,
          "pageUrl": "Shutter Island",
          "rating": "8,1",
          "ratingLevel": "Very good",
          "released": 2010,
          "runTime": "1h 44m",
          "scoresCount": 1002557,
          "srcBgImage": "https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg",
          "srcPosterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg",
          "srcPreviewImage": "https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg",
          "srcPreviewVideo": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          "srcVideo": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
          "starrings": Array [
            "Leonardo DiCaprio",
            "Emily Mortimer",
            "Mark Ruffalo",
          ],
          "title": "Shutter Island",
        }
      }
      id={1}
      onChangeFilmRoute={[MockFunction]}
      route="Overview"
    />
  </BrowserRouter>
</ContextProvider>
`;
